src/util/vecdeque.h

```diff
@@ -0,0 +1,268 @@
+            ResizeDown(size);
+        } else if (size > m_size) {
+            // When growing, first see if we need to allocate more space.
+            if (size > m_capacity) Reallocate(size);
+            while (m_size < size) {
+                std::construct_at(m_buffer + Index(m_size));
+                ++m_size;
+            }
+        }
+    }
+
+    void clear() noexcept { ResizeDown(0); }
+
+    ~VecDeque()
+    {
+        clear();
+        Reallocate(0);
+    }
+
+    VecDeque& operator=(const VecDeque& other)
+    {
+        clear();
+        Reallocate(other.m_size);
+        if constexpr (std::is_trivially_copyable_v<T>) {
+            size_t first_part = other.FirstPart();
+            if (first_part != 0) {
+                std::memcpy(m_buffer, other.m_buffer + other.m_offset, first_part * sizeof(T));
+            }
+            if (first_part != other.m_size) {
+                std::memcpy(m_buffer + first_part, other.m_buffer, (other.m_size - first_part) * sizeof(T));
+            }
+            m_size = other.m_size;
+        } else {
+            while (m_size < other.m_size) {
+                std::construct_at(m_buffer + Index(m_size), other[m_size]);
+                ++m_size;
+            }
+        }
+        return *this;
+    }
+
+    void swap(VecDeque& other) noexcept
+    {
+        std::swap(m_buffer, other.m_buffer);
+        std::swap(m_offset, other.m_offset);
+        std::swap(m_size, other.m_size);
+        std::swap(m_capacity, other.m_capacity);
+    }
+
+    friend void swap(VecDeque& a, VecDeque& b) noexcept { a.swap(b); }
+    VecDeque& operator=(VecDeque&& other) noexcept { swap(other); return *this; }
+    VecDeque(const VecDeque& other) { *this = other; }
+    VecDeque(VecDeque&& other) noexcept { swap(other); }
+
+    bool friend operator==(const VecDeque& a, const VecDeque& b)
+    {
+        if (a.m_size != b.m_size) return false;
+        for (size_t i = 0; i < a.m_size; ++i) {
+            if (a[i] != b[i]) return false;
+        }
+        return true;
+    }
+
+    std::strong_ordering friend operator<=>(const VecDeque& a, const VecDeque& b)
+    {
+        size_t pos_a{0}, pos_b{0};
+        while (pos_a < a.m_size && pos_b < b.m_size) {
+            auto cmp = a[pos_a++] <=> b[pos_b++];
+            if (cmp != 0) return cmp;
+        }
+        return a.m_size <=> b.m_size;
+    }
+
+    void reserve(size_t capacity)
+    {
+        if (capacity > m_capacity) Reallocate(capacity);
+    }
+
+    void shrink_to_fit()
+    {
+        if (m_capacity > m_size) Reallocate(m_size);
+    }
+
+    void push_back(T&& elem)
+    {
+        if (m_size == m_capacity) Reallocate((m_size + 1) * 2);
+        std::construct_at(m_buffer + Index(m_size), std::move(elem));
+        ++m_size;
+    }
+
+    void push_back(const T& elem)
+    {
+        if (m_size == m_capacity) Reallocate((m_size + 1) * 2);
+        std::construct_at(m_buffer + Index(m_size), std::move(elem));
+        ++m_size;
+    }
```

Comment:
Does `std::move(const T&)` make sense here? The `std::construct_at` won't be able to modify `elem` due to its constness, right?